### ISSUE TITLE
Información sobre LSTM

### ISSUE THREAD

**Issue by josprimen** [josprimen/TFG#5]
Buscar información sobre lstm para aprender de qué se trata.

Echarle un vistazo al tutorial proporcionado por el tutor:
- https://machinelearningmastery.com/time-series-prediction-lstm-recurrent-neural-networks-python-keras/

**Comment by josprimen** [comment]
Se asistió a una reunión de alumnos internos del departamento de Inteligencia Artificial en la cual se habó de algunos temas entre los que se encontraba LSTM.

**Comment by josprimen** [comment]
Realización del tutorial de LSTM proporcionado por el profesor.

**Comment by josprimen** [comment]
- LSTM (Long short-term memory) -> Las LSTM son un tipo especial de redes recurrentes. La característica principal de las redes recurrentes es que la información puede persistir introduciendo bucles en el diagrama de la red, por lo que, básicamente, pueden «recordar» estados previos y utilizar esta información para decidir cuál será el siguiente. Esta característica las hace muy adecuadas para manejar series cronológicas. Más info ( https://blog.gft.com/es/2018/11/06/como-usar-redes-neuronales-lstm-en-la-prediccion-de-averias-en-las-maquinas/ )
-La regresión consiste en estudiar la relación entre variables y es útil en problemas de predicción.

**Comment by josprimen** [comment]
Tutorial compañía aérea completado. Se probaron LSTM con diferentes parámetros de Look_Back, Time_Steps, Memory y Return_Sequences. Además se probaron diferentes números de iteraciones.

**Conclusiones**

1. El look_back tiene que casar con la arquitectura de la red. Si no casa puede que un look_back mayor arroje peores resultados.
2. Hacer time_step puede que no nos de mejores resultados o que estos mejoren muy poco, pero la estructura de los datos es más adecuada para la red (en principio).
3. Con memoria las redes funcionan mejor en secuencias largas. En el caso de la compañia aérea no nos mejora los datos. Síntoma de que la red necesita más capas y ser entrenado con más iteraciones probablemente.
4. En redes LSTM podemos obligar a que las capas de este tipo den a la siguiente capa la secuencia. Esto nos ayuda a poder entrenar a la red pese a bloqueos en redes con arquitecturas complejas.

**Comment by josprimen** [comment]
**Pasar datos de no estacionarios a estacionarios facilita en gran medida el trabajo, las redes funcionan mucho mejor con los segundos.**
https://estrategiastrading.com/series-estacionarias/

**Comment by josprimen** [comment]
El **batch_size** sirve para, si tienes 100 ejemplos para entrenar la red y setteas este a 2, coje los 2 primeros ejemplos y entrena la red, después coje los siguientes 2 ejemplos y vuelve a entrenar la red...

https://stats.stackexchange.com/questions/153531/what-is-batch-size-in-neural-network

**Verbose**
By setting verbose 0, 1 or 2 you just say how do you want to 'see' the training progress for each epoch.

verbose=0 will show you nothing (silent)

verbose=1 will show you an animated progress bar like this:

progres_bar

verbose=2 will just mention the number of epoch like this:

enter image description here


https://stackoverflow.com/questions/47902295/what-is-the-use-of-verbose-in-keras-while-validating-the-model




**Shapes**
Shapes are consequences of the model's configuration. Shapes are tuples representing how many elements an array or tensor has in each dimension.

Ex: a shape (30,4,10) means an array or tensor with 3 dimensions, containing 30 elements in the first dimension, 4 in the second and 10 in the third, totaling 30*4*10 = 1200 elements or numbers.

**The input shape**
What flows between layers are tensors. Tensors can be seen as matrices, with shapes.

In Keras, the input layer itself is not a layer, but a tensor. It's the starting tensor you send to the first hidden layer. This tensor must have the same shape as your training data.


https://stackoverflow.com/questions/44747343/keras-input-explanation-input-shape-units-batch-size-dim-etc


**INVERTIR NORMALIZADO**
El tamaño del array al que le vamos a invertir el normalizado debe ser del mismo tamaño que el tamaño del array que normalizamos en origen.

Tener en cuenta que para el normalizado se usan todos los demás datos, por lo que debemos usar los mismos datos que usamos para normalizar a la hora de invertir para que tengan sentido los datos que vamos a obtener.

En el ejemplo de la polución utilizamos un array original con los datos de la polución, lluvia, nieve, etc para predecir los datos de la futura polución. Lo que hemos de hacer es sustituir la primera columna (polución) por los datos predichos por la red y desnormalizar. Simplemente sustituir el dato real por el predicho.

**Comment by josprimen** [comment]
**VECTOR BIAS**
El vector **Bias** es un conjunto de pesos que no requiere entrada. Se corresponde con la salida de una red neuronal artificial cuando el conjunto de entrada se compone en su totalidad de ceros.

Sin el módulo Bias ninguna capa neuronal sería capaz de generar otra cosa que no fueran ceros cuando la entrada se compusiese en su totalidad de ceros.

El Bias al fin y al cabo hace que la variación de pesos, aunque sea pequeña, adquiera importancia dando resultados más distantes. Y hace que si la entrada son 0 no obtengamos el mismo resultado para la salida.

El Bias te permite **desplazar la función de actlivación a la izquierda o la derecha**

**SIN BIAS**
![image](https://user-images.githubusercontent.com/36623141/69917461-b4289680-1466-11ea-85d2-6bdc7a8095e8.png)
![image](https://user-images.githubusercontent.com/36623141/69917568-c2c37d80-1467-11ea-84a3-36525aeb68f2.png)

**CON BIAS**
![image](https://user-images.githubusercontent.com/36623141/69917471-c571a300-1466-11ea-9402-57c9c3ad4517.png)

![image](https://user-images.githubusercontent.com/36623141/69917608-04ecbf00-1468-11ea-9f3e-2e3b9bf60c1e.png)


https://stackoverflow.com/questions/2480650/role-of-bias-in-neural-networks
